### PR TITLE
Project client

### PR DESCRIPTION
**Pull request by lomamech**
Updated part client of unit project and adding the test end to end

@jniles  @DedrickEnc

### CODE REVIEW

File: client/src/js/services/ProjectService.js [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
-function ProjectService($http, util) {
+function ProjectService($http, $translate, util) {
   var service = this;
 
   service.create = create;
   service.read = read;
   service.update = update;
-  service.delete = del;
+  service.del = del;
+  service.readComplete = readComplete;
```
Comment (jniles) [inline client/src/js/services/ProjectService.js]: You can use the [$http](https://docs.angularjs.org/api/ng/service/$http) service to set querystrings for you via a configuration object.  [Here is the relevant documenation](https://docs.angularjs.org/api/ng/service/$http#get).

Instead of creating a whole new method, you should just use `service.read`!  Something like this:

``` js
function read(id, params) {
   var url = (id) ? '/projects/' + id : '/projects';
   return $http.get(url, { params : params });
}

// in your controller
ProjectService.read(null, { complete : 1 })
.then(function (projects) {
  console.log('The complete projects:', projects);
});
```

AngularJS should automatically serialize the parameters into a query string for you.

File: client/src/js/services/ProjectService.js [hunk 2]
```diff
@@ -23,13 +24,24 @@ function ProjectService($http, util) {
...
+  }
+
   function update(id, project) {
     return $http.put('/projects/' + id, project)
     .then(util.unwrapHttpResponse);
   }
 
   function del(id) {
+
+    var result = confirm($translate.instant('PROJECT.CONFIRM'));
```
Comment (jniles) [inline client/src/js/services/ProjectService.js]: You should use `$window.confirm` so that we can test it in the future.  [Relevant documentation](https://docs.angularjs.org/api/ng/service/$window).

File: client/src/js/services/SnisService.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  .factory('SnisService', SnisService);
+  
+SnisService.$inject = ['$http', 'util'];
+
+function SnisService($http, util) {  
+  var service = {};
+
+  service.readSnisZs = SnisZs;
+
+  function SnisZs() {
```
Comment (jniles) [inline client/src/js/services/SnisService.js]: A better name for this function would be `healthZones()`.  You can export it as `SnisService.healthZones()` as well.  A better URL would be `/snis/health_zones`.  The 'get_' word is implicit in the HTTP verb "GET".

File: client/src/partials/projects/project.js [hunk 4]
```diff
@@ -1,117 +1,118 @@
...
+    // load Enterprises
+    Enterprises.read().then(function (data) {
+      vm.enterprises = data;
+    }).catch(handler);
+
+    SnisService.readSnisZs().then(function (data) {
+      data.forEach(function (l) {
+        l.format = l.zone + ' - ' + l.territoire + ' (' + l.province + ')';
+      });
+      vm.snis_zs = data;
```
Comment (jniles) [inline client/src/partials/projects/project.js]: Why not just call this `zones`?  It's much more readable.

File: client/src/partials/projects/project.js [hunk 5]
```diff
@@ -1,117 +1,118 @@
...
-    }
-  };
-
-  function handleErrors(err) {
-    messenger.danger('An Error occured : ' + err);
+function ProjectController(Projects, Enterprises, SnisService, StateFactory) {
+  var vm = this;
+
+  vm.enterprises = [];
+  vm.state = new StateFactory();
```
Comment (jniles) [inline client/src/partials/projects/project.js:13]: In the future (not this PR), you probably shouldn't use `StateFactory()`.  It was an experiment that I created last Fall, and failed to yield a usable service.  The API is terrible and it causes more code confusion than clarity.

@sfount is investigating [ui-router](https://github.com/angular-ui/ui-router) to try and find a better way to manage state in the application.  If you can find a way to encapsulate state better, I'd be interested in seeing a proposal!

File: client/test/e2e/projects/projects.spec.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+describe('The Projects Module', function () {
+  // shared methods
+  var path = '#/projects';
+  var PROJECT = {
+    name : 'PROJECT END 2 END',
+    abbr : 'P e2e'
+  };
+
+  //To obtain the rank of a random element to the project list
```
Comment (jniles) [inline client/test/e2e/projects/projects.spec.js:18]: Cool!

File: client/test/e2e/shared/FormUtils.js [hunk 7]
```diff
@@ -101,6 +101,10 @@ module.exports = {
     return element.all(by.model(model)).get(n).click();
   },
 
+  // get a checked button by its position and click
+  checked : function checked(model, n) {
```
Comment (jniles) [inline client/test/e2e/shared/FormUtils.js]: Is this ever used in your code?  Do you have a use case in mind?

File: server/controllers/medical/snis.js [hunk 8]
```diff
@@ -290,11 +290,22 @@ function getReport (req, res) {
 	}
 }
 
+function get_snis_zs (req, res) {
```
Comment (jniles) [inline server/controllers/medical/snis.js]: A better name for this would be `function healthZones()`.

Why are you not using `next()`?!?

File: server/controllers/medical/snis.js [hunk 9]
```diff
@@ -290,11 +290,22 @@ function getReport (req, res) {
...
+
+  var sql = 'SELECT id, zone, territoire, province FROM mod_snis_zs';
+  db.exec(sql, function (err, rows) {
```
Comment (jniles) [inline server/controllers/medical/snis.js]: There is no error handling here.

File: server/controllers/medical/snis.js [hunk 10]
```diff
@@ -290,11 +290,22 @@ function getReport (req, res) {
...
+    if (err) { throw err; }
+    res.send(rows);
```
Comment (jniles) [inline server/controllers/medical/snis.js]: What HTTP status should this have?

**Comment by jniles** [comment]
@lomamech, I've completed an initial review.  Somethings to remember:
1. All of the server routes **must** have `req, res, next` as parameters.  Using `next()` is the best practice for handling errors in your routes.  In your `get_snis_zs` route, you do not have any error handling at all.
2. [According to Travis](https://travis-ci.org/IMA-WorldHealth/bhima-2.X/jobs/106481776), you have an error in your SQL syntax.  (In MySQL, you cannot have comments like "--comment".... there must be a space in between the "--" and the "comment").  Did you actually build the test database before submitting this PR?

Thank you for using the "@jniles" syntax.  Let me know when you've look into these errors.

**Comment by lomamech** [comment]
@jniles I'm looking at thes errors and I start simultaneously correction

**Comment by lomamech** [comment]
@jniles @DedrickEnc I just correct errors but also to make improvements,

**Comment by jniles** [comment]
@lomamech, this is good to see.  Your integration tests do not pass.  [Here are your results](https://travis-ci.org/IMA-WorldHealth/bhima-2.X/jobs/106688795).

I'll review the code when the tests pass.

File: client/src/js/services/ProjectService.js [hunk 11]
```diff
@@ -1,26 +1,24 @@
...
 
-function ProjectService($http, util) {
+function ProjectService($http, $translate, $window, util) {
   var service = this;
 
   service.create = create;
   service.read = read;
   service.update = update;
-  service.delete = del;
+  service.del = del;
```
Comment (jniles) [inline client/src/js/services/ProjectService.js]: Why did you change this?

File: client/src/js/services/SnisService.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
+angular.module('bhima.services')
+  .factory('SnisService', SnisService);
```
Comment (jniles) [inline client/src/js/services/SnisService.js]: Since this is in the module "bhima.services" it's better to use a service.  They are almost identical.  See #53.

File: client/src/partials/projects/project.js [hunk 13]
```diff
@@ -1,117 +1,119 @@
...
+    console.error(error);
+    vm.state.error();
   }
 
-  function load(models) {
-    for (var k in models) { $scope[k] = models[k]; }
+  // fired on startup
+  function startup() {
+    // load Projects
+    Projects.read(null, { complete : 1 })
```
Comment (jniles) [inline client/src/partials/projects/project.js]: Much better API!

... you already have a function `refreshProjects()` later in this exact same controller.  Could you refactor this code into a single function to load data from the server?  It will be easier to maintain if the API ever changes.

File: client/src/partials/projects/project.js [hunk 14]
```diff
@@ -1,117 +1,119 @@
...
+    vm.state.reset();
+    vm.view = 'default';
+  }
+
+  function create() {
+    vm.view = 'create';
+    vm.project = {};
+  }
+
+  // Load a Project from the server
```
Comment (jniles) [inline client/src/partials/projects/project.js]: The comment is not accurate.  it doesn't load any data from the server.  It simply binds data to the view.  Also, `data` here is an ID, isn't `id` be a better variable name?

... why does this need to be in a separate function?

File: client/src/partials/projects/project.js [hunk 15]
```diff
@@ -1,117 +1,119 @@
...
-      ['project.enterprise_id=' + project.enterprise_id];
-    validate.process(dependencies)
-    .then(load, handleErrors);
-  });
+  }
+
+  // form submission
+  function submit(invalid) {
+    if (invalid) { return; }
+    vm.project.locked = (vm.project.locked)?1:0;
```
Comment (jniles) [inline client/src/partials/projects/project.js]: Why do you have to do this?  What is `vm.project.locked`, if not a 0 or 1?

File: server/controllers/medical/snis.js [hunk 16]
```diff
@@ -290,11 +290,24 @@ function getReport (req, res) {
 	}
 }
 
+function health_zones (req, res, next) {
```
Comment (jniles) [inline server/controllers/medical/snis.js]: Please use camelCase instead of snake_case for JavaScript variables, functions, and objects.

File: client/src/js/services/ProjectService.js (hunk 11 above)
Comment (lomamech) [inline client/src/js/services/ProjectService.js]: @jniles  just to not have a big difference with the other function

Comment (jniles) [inline client/src/js/services/ProjectService.js]: That's a good intention, but the reason we cannot call the function "delete" is because `delete` is a javascript reserved word.  To make the API clear, the API should still be called `service.delete`.

File: client/src/partials/projects/project.js (hunk 15 above)
Comment (lomamech) [inline client/src/partials/projects/project.js]: @jniles Just because the element form the checkbox returns true instead of 1 and 0, and the test clip 1 if locked is checked and 0 otherwise

Comment (jniles) [inline client/src/partials/projects/project.js]: Can you make the form return the value you want using `ng-true-value` and `ng-false-value`?  [Reference Documentation](https://docs.angularjs.org/api/ng/input/input%5Bcheckbox%5D).

File: client/src/js/services/SnisService.js [hunk 17]
```diff
@@ -0,0 +1,18 @@
+angular.module('bhima.services')
+.service('SnisService', SnisService);
+  
+SnisService.$inject = ['$http', 'util'];
+
+function SnisService($http, util) {  
+  var service = {};
```
Comment (jniles) [inline client/src/js/services/SnisService.js]: `var service = this`.

**Comment by jniles** [comment]
@lomamech, the end to end tests pass for me on master, but do not pass on your branch.  [I've uploaded your results](https://github.com/IMA-WorldHealth/bhima-2.X/files/115647/protractor-errors.txt).

**Please, please, run end to end tests and integration tests before submitting a PR.**

**Comment by lomamech** [comment]
@jniles   end to end tests and integration tests run very well

**Comment by jniles** [comment]
[14 passing, 6 failing end to end tests](https://github.com/IMA-WorldHealth/bhima-2.X/files/115873/protractor-errors.txt).

**Comment by lomamech** [comment]
@jniles 6 test end to end had fails because the data format sended by ProjectService

File: client/src/partials/projects/projects.html [hunk 18]
```diff
@@ -1,21 +1,17 @@
...
-    </button>
-  </div>
-</nav>
+<div class="flex-util">
+  <button class="btn btn-sm btn-default" id="create" ng-click="ProjectCtrl.create()" data-method="create">
+    <span class="glyphicon glyphicon-plus-sign"></span> {{ 'PROJECT.NEW' | translate }}
+  </button>
+</div>
 
 <div class="report visible-print">
```
Comment (jniles) [inline client/src/partials/projects/projects.html]: You can delete this entire section (`<div class="report visible-print>"`) - it is no longer used and is bad practice. It also slows down the loading of the module.

File: client/src/partials/projects/projects.html [hunk 19]
```diff
@@ -44,120 +40,188 @@
...
 </div>
 
-<main class="hidden-print">
-  <div class="row">
-
-    <div class="col-xs-6">
-      <div class="panel panel-primary">
-        <div class="panel-heading">
-          {{ 'PROJECT.EXISTING' | translate }}
+<main class="flex-content">
```
Comment (jniles) [inline client/src/partials/projects/projects.html]: To be consistent with @sfount's new style, this element should be a `<div>` not a `<main>`.  The main element may still inherit old bhima-1.X styles.

File: client/src/partials/projects/projects.html [hunk 20]
```diff
@@ -44,120 +40,188 @@
...
+              <tr>
+                <th>{{ 'COLUMNS.ID' | translate }}</th>
+                <th>{{ 'COLUMNS.NAME' | translate }}</th>
+                <th>{{ 'COLUMNS.ABBR' | translate }}</th>
+                <th>{{ 'COLUMNS.ENTERPRISE_ID' | translate }}</th>
+                <th colspan="2">{{ 'COLUMNS.ACTIONS' | translate }}</th>
+              </tr>
+            </thead>
+            <tbody>
+              <tr ng-repeat="project in ProjectCtrl.projects | orderBy:'name'" ng-class="{'btn-primary' : editProject.id==project.id }">
```
Comment (jniles) [inline client/src/partials/projects/projects.html]: For performance reasons, could you include a "track by project.id" in this `ng-repeat` loop?  [Relevant Documentation](https://docs.angularjs.org/api/ng/directive/ngRepeat).

File: client/src/partials/projects/projects.html [hunk 21]
```diff
@@ -44,120 +40,188 @@
...
+                <td>{{ project.id }}</td>
+                <td>{{ project.name }}</td>
+                <td>{{ project.abbr }}</td>
+                <td>{{ project.enterprise_id }}</td>
+                <td><a ng-click="ProjectCtrl.update(project)" id="project-upd-{{ $index + 1}}"><span class="glyphicon glyphicon-pencil"></span></a></td>
+                <td><a ng-click="ProjectCtrl.del(project)" id="project-del-{{ $index + 1 }}" disabled class="danger"><span class="glyphicon glyphicon-trash"></span></a></td>
```
Comment (jniles) [inline client/src/partials/projects/projects.html]: Could you remove the disabled tag from this element?

File: client/src/partials/projects/projects.html [hunk 22]
```diff
@@ -44,120 +40,188 @@
...
+                <th colspan="2">{{ 'COLUMNS.ACTIONS' | translate }}</th>
+              </tr>
+            </thead>
+            <tbody>
+              <tr ng-repeat="project in ProjectCtrl.projects | orderBy:'name'" ng-class="{'btn-primary' : editProject.id==project.id }">
+                <td>{{ project.id }}</td>
+                <td>{{ project.name }}</td>
+                <td>{{ project.abbr }}</td>
+                <td>{{ project.enterprise_id }}</td>
+                <td><a ng-click="ProjectCtrl.update(project)" id="project-upd-{{ $index + 1}}"><span class="glyphicon glyphicon-pencil"></span></a></td>
```
Comment (jniles) [inline client/src/partials/projects/projects.html]: Both these `<a>` button elements should have `class="btn btn-xs"` to distinguish them from links for our users.

File: client/src/partials/projects/projects.html (hunk 21 above)
Comment (jniles) [inline client/src/partials/projects/projects.html]: Also, clicking "cancel" on `window.confirm()` gives me errors in my browser.

File: client/src/partials/projects/projects.html [hunk 23]
```diff
@@ -44,120 +40,188 @@
...
+                  <select class="form-control" name="zs_id" ng-model="ProjectCtrl.project.zs_id" id="bhima-project-zs" ng-options="zs.id as zs.format for zs in ProjectCtrl.zones | orderBy:'zone'" required>
+                    <option value="" disabled> -- {{ 'SELECT.ZS' | translate }} -- </option>
+                  </select>
+                </div>
+              </div>
+              <!-- end zone de sante for snis -->
+              <div class="form-group">
+                <label class="col-md-3 control-label">{{ 'COLUMNS.LOCKED' | translate }}</label>
+                <div class="col-md-9">                
+                  <input type="checkbox" id="locked" ng-true-value="1" ng-false-value="0" name="bhima-project-locked" ng-model="ProjectCtrl.project.locked">
```
Comment (jniles) [inline client/src/partials/projects/projects.html]: The proper way to make a bootstrap checkbox input is to use the "checkbox" class.  [Here is the documentation](http://getbootstrap.com/css/#forms-horizontal).  You should be using the "checkbox" class on a `<div>` and wrapping the `<label>` element around the checkbox.

File: client/src/partials/projects/projects.html [hunk 24]
```diff
@@ -44,120 +40,188 @@
...
+          </div>        
+        </div>
+
+        <div class="panel panel-primary" ng-switch-when="update">
+          <div class="panel-heading">{{ 'PROJECT.EDIT' | translate }}</div>
+          <div class="panel-body">
+            <form class="form-horizontal" name="UpdateForm" ng-submit="ProjectCtrl.submit(UpdateForm.$invalid)" novalidate>
+
+              <div class="form-group" ng-class="{ 'has-error' : UpdateForm.$submitted && UpdateForm.name.$invalid }">
+                <label class="col-xs-3 control-label required" for="bhima-project-name">{{ 'COLUMNS.NAME' | translate }}</label>
```
Comment (jniles) [inline client/src/partials/projects/projects.html]: The `required` class no longer does anything.  You should no longer use it.  Please remove it from any `<label>` element in this file that uses it.

File: client/test/e2e/projects/projects.spec.js [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+    element(by.id('project-del-' + DELETE_SUCCESS )).click();
+    browser.switchTo().alert().accept();
+    expect(element(by.id('delete_success')).isPresent()).to.eventually.be.true;
+  });
+
+  it('No way to delete a project', function () {
+    element(by.id('project-del-' + DELETE_ERROR )).click();
+    browser.switchTo().alert().accept();
+    expect(element(by.id('delete_error')).isPresent()).to.eventually.be.true;
+  });
```
Comment (jniles) [inline client/test/e2e/projects/projects.spec.js:103]: Could you include a test for "canceling a delete does not delete the project"?  I discovered a console error when I clicked "cancel" on the $window.confirm() input.

**Comment by jniles** [comment]
@lomamech, these are very nice tests.  I'm glad that the ProjectService API change was caught, that could have been a painful bug in the future.

I've made some comments, mostly on HTML and CSS style.  Since we are now using bootstrap rules for styling, it is important that we abide by bootstrap standards.  There is also a whole section that can be deleted - it was useful before we learned how to create PDF reports.

There is one test to add based on a bug that I found in the $window.confirm() alert box.  That bug should be fixed before this PR can be merged.

Overall, this code and PR is looking very good.  Once it is bug free and up to standards, it will be a valuable contribution.

**Comment by lomamech** [comment]
@jniles did you analyze the latest changes

File: client/src/partials/projects/projects.html (hunk 18 above)
Comment (jniles) [inline client/src/partials/projects/projects.html]: This is still not fixed.  Please delete this entire `<div>`.

**Comment by jniles** [comment]
You must let me know when to analyze changes again by leaving a comment.  Otherwise, I do not know if you will add more or not.

You have one more change to make, then this can be merged in, assuming it passes tests.  There's  a lot of unused HTML that should be deleted.

**Comment by lomamech** [comment]
@jniles I just did it

**Comment by jniles** [comment]
LGTM.
